zbus_address/src/address.rs

```diff
@@ -0,0 +1,233 @@
+use std::{borrow::Cow, collections::HashSet, fmt};
+
+use super::{
+    percent::{decode_percents, decode_percents_str, Encodable},
+    transport::Transport,
+};
+use crate::{Error, Result};
+
+/// A bus address.
+///
+/// Example:
+/// ```
+/// use zbus_address::DBusAddr;
+///
+/// let addr: DBusAddr = "unix:path=/tmp/dbus.sock".try_into().unwrap();
+/// ```
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct DBusAddr<'a> {
+    pub(super) addr: Cow<'a, str>,
+}
+
+impl<'a> DBusAddr<'a> {
+    /// The connection GUID if any.
+    pub fn guid(&self) -> Option<Cow<'_, str>> {
+        self.get_string("guid").and_then(|res| res.ok())
+    }
+
+    /// Transport connection details
+    pub fn transport(&self) -> Result<Transport<'_>> {
+        self.try_into()
+    }
+
+    fn validate(&self) -> Result<()> {
+        self.transport()?;
+        let mut set = HashSet::new();
```

Comment:
So we go through the trouble of avoiding allocations in the `Address` (even though the 99.99% cases we can't avoid allocation) only to allocate for validation?

I'm not too sure anymore if lifetimes are really worth the effort here.

Reply:
validation can be made optional. alternatively, `validate()` could be improved to avoid allocation. I can add a fixme?

Reply:
> validation can be made optional. alternatively, `validate()` could be improved to avoid allocation. I can add a fixme?

None of that addresses my concerns. Sure, if validation can be allocation free, it'd make me feel much better about the API designed to allow reference types. However, the main issue I see is that for most uses (system and session bus), we'll be allocating and that beats the whole point of lifetimes in types.

However, maybe there is a use case that's quite common where users would be able to avoid allocation, that I'm missing here?

Reply:
"or most uses (system and session bus), we'll be allocating" why?

Reply:
because `std::env::var` allocates. I guess that is not typically true for system but session bus address is almost always taken from the env.